data/jsonic1_commit.py

```diff
@@ -0,0 +1,70 @@
+"""
+Grabs data from the "FAQ Content" CSV and turns it into nice JSON: a main faq object containing an array of themed Section objects, each Section Object in turn holding an array of Question objects consisting of question, answer and related link/s, as follows:
+    [
+        {
+          "Q": "What are the symptoms of COVID-19?",
+          "A": "Symptoms of COVID-19 include coughing and shortness of breath. Additionally, a person showing two or more of the following symptoms may have the virus: fever, repeated shaking with chills, muscle pain, headache, sore throat, new loss of taste or smell.",
+          "link": "https://www.cdc.gov/coronavirus/2019-ncov/symptoms-testing/symptoms.html"
+        },
+        
+    .
+    .
+    .
+    ]
+    etc
+In addition to its array of Question objects, each Section displays a title and "last updated" date value. For the Python stage of the transition from CSV to JSON we create a hierarchy of nested dictionaries and lists equivalent to the JSON objects and arrays.
+"""
+
+import requests
+import json
+import csv
+from io import StringIO
+import datetime
+
+# Link to "FAQ Content" CSV
+link = "https://docs.google.com/spreadsheets/d/1_wBXS62S5oBQrwetGc8_-dFvDjEmNqzqHwUeP-DzkYs/export?format=csv&id=1_wBXS62S5oBQrwetGc8_-dFvDjEmNqzqHwUeP-DzkYs&gid=1318925039"
+
+# Get the data
+response = requests.get(link)
+response.raise_for_status()
+
+# Turn text from CSV stream into a file object csv.reader can read
+csv_stream = StringIO(response.text)
+# Create a reader to iterate over the CSV stream file object
+reader = csv.reader(csv_stream)
+
+# ISO8601 format date, to update each time the program runs
+date = datetime.datetime.now().strftime("%Y-%m-%d")
+
+# Create the main FAQ dictionary
+# Create a list for our Sections
+# Place the Sections list inside the main dictionary
+faq_dict = {}
+sections_list = []
+faq_dict["faqItems"] = sections_list
+
+# Get data from the reader object
+# Arrange Sections and Questions in a JSON-friendly way
+for row in reader:
+    rowtype = row[0]   # CSV column delineating Section, Question, Answer or Link
+    rowval = row[1]   # Values for the above
```

Comment:
Hi Rob. Love this construction, which I've now got my head around I think. One quick question: some of the other columns in each row can contain things like the initials of whoever approved them (usually me or Niko) or suggested links that weren't used etc, or occasionally notes. Will this interfere with the output?

Reply:
Or does it automatically assign rowtype and rowval to the first two columns containing data?

Reply:
>> Will [extra columns] interfere with the output?
>
> Or does it automatically assign rowtype and rowval to the first two columns containing data?

Right. And the extra columns *would* be a problem, except we solve that by using the asterisk in `*_`. Let’s say you had a CSV like:

| Layout | Final Content | Submitted by | Reviewed by | New Content |
| -- | -- | -- | -- | -- |
| Q  | What are the symptoms of COVID-19? | Niko | Andrew | Whatever |

Which `csv.reader` will surface as the list:

```py
['Q', 'What are the symptoms of COVID-19?', 'Niko', 'Andrew', 'Whatever']
```

So what you get with “destructuring assignment” (when you assign multiple variables this way) is:

```py
row = ['Q', 'What are the symptoms of COVID-19?', 'Niko', 'Andrew', 'Whatever']
rowtype, rowval, *_ = row
rowtype == 'Q'  # The first item in `row`
rowval  == 'What are the symptoms of COVID-19?'  # The second item in `row`
_       == ['Niko', 'Andrew', 'Whatever']  # The rest of the items in `row`
```

The asterisk is important, because it’s what makes `_` into the *rest* of the items, rather than just the third item. If we didn’t have it, Python would actually raise an exception because there are more values than variables you provided to assign them to:

```py
row = ['Q', 'What are the symptoms of COVID-19?', 'Niko', 'Andrew', 'Whatever']
# This generates an exception:
rowtype, rowval, _ = row

# Python is essentially doing this:
rowtype == 'Q'  # The first item in `row`
rowval  == 'What are the symptoms of COVID-19?'  # The second item in `row`
_       == 'Niko'  # The third item in `row`
# OH NO, there's no variable(s) to put the 4th and 5th items in! Raise an error!
```

If you knew every row reliably had exactly three items, you could drop the asterisk and it would be safe (or drop the `_` altogether if each row only had two items).

Reply:
Actually, as a better alternative, have you looked into [`csv.DictReader`](https://docs.python.org/3.8/library/csv.html#csv.DictReader)? I’m not sure why I didn’t suggest it before, but it makes each row a dict, where the keys are based on the CSV’s column headers. So instead of:

```py
for row in reader:
    rowtype, rowval, *_ = row
    if rowtype == "Section Head":
        ...
```

You’d have:

```py
for row in reader:
    # The keys in `row` are based on the column headers
    # (i.e. the values in the first row of the CSV)
    if row['Layout'] == "Section Head":
        ...
```